### PR TITLE
RavenDB gateway implementation

### PR DESCRIPTION
**Pull request by mauroservienti**
Replaces #1

### CODE REVIEW

File: src/NServiceBus.Gateway.RavenDB.AcceptanceTests/GatewayTestSuiteConstraints.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+                documentStore.Maintenance.Server.Send(new CreateDatabaseOperation(new DatabaseRecord(databaseName)));
+
+                return documentStore;
+            });
+
+            var gatewaySettings = configuration.Gateway(ravenGatewayDeduplicationConfiguration);
+            configuration.GetSettings().Set(gatewaySettings);
+
+            return Task.FromResult(false);
```
Comment (DavidBoike) [inline src/NServiceBus.Gateway.RavenDB.AcceptanceTests/GatewayTestSuiteConstraints.cs]: Doesn't Raven 4 force you into net461 minimum?
```suggestion
            return Task.CompletedTask;
```

File: src/NServiceBus.Gateway.RavenDB/NServiceBus.Gateway.RavenDB.csproj [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  <PropertyGroup>
+    <TargetFrameworks>net461;netstandard2.0</TargetFrameworks>
+    <SignAssembly>true</SignAssembly>
+    <AssemblyOriginatorKeyFile>$(SolutionDir)NServiceBus.snk</AssemblyOriginatorKeyFile>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="NServiceBus" Version="[7.2.1, 8.0.0)" />
+    <PackageReference Include="NServiceBus.Gateway" Version="[3.1.0, 4.0.0)" />
+    <PackageReference Include="Particular.Packaging" Version="0.5.0" PrivateAssets="All" />
```
Comment (DavidBoike) [inline src/NServiceBus.Gateway.RavenDB/NServiceBus.Gateway.RavenDB.csproj]: I forgot ConfigureAwaits until last minute.
```suggestion
    <PackageReference Include="Particular.Packaging" Version="0.5.0" PrivateAssets="All" />
    <PackageReference Include="Particular.CodeRules" Version="0.2.1" PrivateAssets="All" />
```

File: src/NServiceBus.Gateway.RavenDB/RavenGatewayDeduplicationConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            return new RavenGatewayDeduplicationStorage(documentStore, DeduplicationDataTimeToLive);
+        }
+
+        static void EnableExpirationFeature(IDocumentStore documentStore, long frequencyToRunDeduplicationDataCleanup)
+        {
+            documentStore.Maintenance.Send(new ConfigureExpirationOperation(new ExpirationConfiguration
+            {
+                Disabled = false,
+                DeleteFrequencyInSec = frequencyToRunDeduplicationDataCleanup
+            }));
```
Comment (DavidBoike) [inline src/NServiceBus.Gateway.RavenDB/RavenGatewayDeduplicationConfiguration.cs]: This becomes a database-wide setting?

Comment (mauroservienti) [reply src/NServiceBus.Gateway.RavenDB/RavenGatewayDeduplicationConfiguration.cs]: unfortunately yes, that's the reason `frequencyToRunDeduplicationDataCleanup` is exposed to users. If the Expiration feature is already enabled for other reasons we cannot change its settings.

File: src/NServiceBus.Gateway.RavenDB/NServiceBus.Gateway.RavenDB.csproj (hunk 2 above)
Comment (mauroservienti) [reply src/NServiceBus.Gateway.RavenDB/NServiceBus.Gateway.RavenDB.csproj]: Addressed

File: src/NServiceBus.Gateway.RavenDB.AcceptanceTests/GatewayTestSuiteConstraints.cs (hunk 1 above)
Comment (mauroservienti) [reply src/NServiceBus.Gateway.RavenDB.AcceptanceTests/GatewayTestSuiteConstraints.cs]: Addressed

**Pull request merged**